### PR TITLE
Fix#175

### LINKED ISSUE #175: Optimisation: Detect duplicate RPC/Command
If the test suite contains multiple instance of the same command/rpc, the tool should avoid running it again and again in the same execution

### PR DESCRIPTION
**Pull request by awasthiy**
Fix #175 
Optimisation: Detect duplicate RPC/Command

If the test suite contains multiple instance of the same command/rpc, Tool avoid running it again and again in the same execution whether it is in same test file or in different test files

config.yml file
```
hosts:
  - device: 10.209.16.204 
    username : demo
    passwd: demo123
tests:
  - test_is_in.yml
  - test_not_in.yml
  - test_is_gt.yml

```
test_is_in.yml

```
tests_include:
  - check_show_interfaces_terse

check_show_interfaces_terse:
  - command: show interfaces terse lo*
  - iterate: 
      xpath: //physical-interface[normalize-space(name) = "lo0"]
      tests:
        - is-in: oper-status, downoo, up
          info: "Test Succeeded!! Physical operational status is-in downoo-up, it is: <{{post['oper-status']}}> with admin status <{{post['admin-status']}}>"
          err: "Test Failed!!! Physical operational status is not in downoo-up, it is: <{{post['oper-status']}}> with admin status <{{post['admin-status']}}> "

```
test_not_in.yml
```
tests_include:
  - check_show_interfaces_terse_test

check_show_interfaces_terse_test:
  - command: show interfaces terse lo*
  - iterate: 
      xpath: //physical-interface[normalize-space(name) = "lo0"]
      tests:
        - not-in: oper-status, downward, upward
          info: "Test Succeeded!! Physical operational status is not in  downward-upward, it is: <{{post['oper-status']}}> with admin status <{{post['admin-status']}}>"
          err: "Test Failed!!! Physical operational status is in downward-upward, it is: <{{post['oper-status']}}> with admin status <{{post['admin-status']}}> "

```
test_is_gt.yml
```
tests_include:
  - check_chassis_fpc
  - check_chassis_fpc_2   

check_chassis_fpc:
  - command: show chassis fpc
  - iterate:
      id: ./memory-dram-size
      xpath: //fpc[normalize-space(slot) = "0"]
      tests:
        - is-gt: cpu-total, 2
          info: "Test Succeeded!!, cpu-total is greater than 2, it is  <{{post['cpu-total']}}> with id {{id_0}} and memory-heap-utilization: <{{post['memory-heap-utilization']}}>"
          err: "Test Failed!!!, cpu-total is not greater than 2, it is <{{post['cpu-total']}}> with id {{id_0}} and memory-heap-utilization: <{{post['memory-heap-utilization']}}> "

check_chassis_fpc_2:
  - command: show chassis fpc
  - iterate: 
      xpath: fpc[normalize-space(slot) = "0"]
      tests:
        - in-range: memory-heap-utilization, 5, 40
          info: "Test Succeeded!! memory heap utilisation of the FPCs is within the range of 5-40, it is <{{post['memory-heap-utilization']}}> with temperature: <{{post['temperature']}}>"
          err: "Test Failed!! memory heap utilisation of the FPCs is not in range of 5-40, it is <{{post['memory-heap-utilization']}}> with temperature: <{{post['temperature']}}>"

```

Following output occured :

```
Connecting to device 10.221.130.68 ................
Taking snapshot of COMMAND: show interfaces terse lo* 
Taking snapshot of COMMAND: show chassis fpc 
*************************** Device: 10.221.130.68 ***************************
Tests Included: check_show_interfaces_terse 
********************* Command: show interfaces terse lo* *********************
PASS | All "oper-status" is in list ['downoo', 'up']  [ 1 matched ]
*************************** Device: 10.221.130.68 ***************************
Tests Included: check_show_interfaces_terse_test 
********************* Command: show interfaces terse lo* *********************
PASS | All "oper-status" is not in list ['downward', 'upward']  [ 1 matched ]
*************************** Device: 10.221.130.68 ***************************
Tests Included: check_chassis_fpc 
************************* Command: show chassis fpc *************************
PASS | All "cpu-total" is greater than 2" [ 1 matched ]
Tests Included: check_chassis_fpc_2 
************************* Command: show chassis fpc *************************
PASS | All "memory-heap-utilization" is in range "5.000000 - 40.000000" [ 1 matched ]
------------------------------- Final Result!! -------------------------------
check_show_interfaces_terse : Passed
check_chassis_fpc_2 : Passed
check_chassis_fpc : Passed
check_show_interfaces_terse_test : Passed
Total No of tests passed: 4
Total No of tests failed: 0 
Overall Tests passed!!! 

```

### CODE REVIEW

File: samples/config_diff.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ hosts:
     username: demo 
     passwd: demo123
 tests:
-  - test_diff.yml 
+  - test_diff.yml
```
Comment (eeishaan) [inline samples/config_diff.yml]: Unnecessary commit

File: tests/unit/configs/is-gt_ignore-null_fail.yml [hunk 2]
```diff
@@ -9,4 +9,3 @@ check_chassis_fpc:
         - is-gt: cpu-total, 1
           info: "Test Succeeded!! cpu total is greater than 1 and its value is {{post['cpu-total']}}"
           err: "Test Failed!!! cpu total is less than 1 and its value is {{post['cpu-total']}}"
-
```
Comment (eeishaan) [inline tests/unit/configs/is-gt_ignore-null_fail.yml]: Unnecessary line deletion, revert back to original form

File: lib/jnpr/jsnapy/snap.py [hunk 3]
```diff
@@ -135,19 +135,21 @@ def run_cmd(self, test_file, t, formats, dev, output_file, hostname, db):
         self.logger_snap.debug(colorama.Fore.BLUE +
                                "Tests Included: %s " %t,
                                extra=self.log_detail)
-        self.logger_snap.info(
-            colorama.Fore.BLUE +
-            "Taking snapshot of COMMAND: %s " %
-            command,
-            extra=self.log_detail)
+        if command not in command_list:
```
Comment (eeishaan) [inline lib/jnpr/jsnapy/snap.py]: This is not the correct way to implement unique command execution. This only stops the printing. In the `else` scenario, the snapshot is still taken and the optimization is not at all achieved

File: lib/jnpr/jsnapy/snap.py [hunk 4]
```diff
@@ -214,10 +216,12 @@ def run_rpc(self, test_file, t, formats, dev, output_file, hostname, db):
         self.logger_snap.debug(colorama.Fore.BLUE +
                                "Tests Included : %s " %t,
                                extra=self.log_detail)
-        self.logger_snap.info(colorama.Fore.BLUE +
-                              "Taking snapshot of RPC: %s" %
-                              rpc,
-                              extra=self.log_detail)
+        if rpc not in rpc_list:
```
Comment (eeishaan) [inline lib/jnpr/jsnapy/snap.py]: Here also the same problem as above is there. In the `else` scenario the snapshot is still taken. Stop the snapshot procedure in the else scenario.

File: lib/jnpr/jsnapy/jsnapy.py [hunk 5]
```diff
@@ -373,8 +373,10 @@ def generate_rpc_reply(self, dev, output_file, hostname, config_data):
                     tfile, extra=self.log_detail)
 
         g = Parser()
+        command_list = []
```
Comment (eeishaan) [inline lib/jnpr/jsnapy/jsnapy.py]: Add a comment here to make it explicit that command_list and rpc_list persist across test files and as they are passed by reference to generate_reply they help achieve command and rpc uniqueness

File: tests/unit/test_snap.py [hunk 6]
```diff
@@ -299,6 +299,64 @@ def test_rpc_6(self, mock_etree, mock_parse):
...
+                self.db,
+                command_list,
+                rpc_list)
+            self.assertEqual(command_list,['show bgp neighbor'])
+            self.assertEqual(rpc_list,['get-bgp-neighbor-information'])
+        dev.close()
+    @patch('jnpr.junos.rpcmeta._RpcMetaExec.__getattr__')
+    @patch('jnpr.junos.device.Device')
+    @patch('jnpr.jsnapy.snap.etree')
+    def test_8(self, mock_etree, mock_dev, mock_rpc):
```
Comment (eeishaan) [inline tests/unit/test_snap.py]: inconsistent function name

File: tests/unit/test_snap.py [hunk 7]
```diff
@@ -299,6 +299,64 @@ def test_rpc_6(self, mock_etree, mock_parse):
             c = mock_log.call_args_list[0]
             self.assertNotEqual(c[0][0].find("ERROR occurred"), -1)
 
+
```
Comment (eeishaan) [inline tests/unit/test_snap.py]: Add test case for scenario where there are multiple test files and duplicate command/rpc are distributed among the files

**Comment by eeishaan** [comment]
Refer #230 . @awasthiy does this pull request handle this scenario? 
If kwargs are present with an RPC it's essentially a different RPC.

**Comment by sidhujasminder** [comment]
No Longer valid, as the issue is closed.
